### ISSUE TITLE
Tiles appear on incorrect monitor when moving window with mouse

### ISSUE THREAD

**Issue by efogdev** [domferr/tilingshell#194]
![opt](https://github.com/user-attachments/assets/2e617786-d496-4536-871e-80c274674fd7)

**Information (please complete the following):**
 - Tiling Shell version: 15 (test build)
 - GNOME version: 47.1

**Additional context**
```
Nov 17 23:56:40 efog-laptop gnome-shell[320243]: [tilingshell] [extension] building a tiling manager for each monitor
Nov 17 23:56:40 efog-laptop gnome-shell[320243]: [tilingshell] [TilingManager 0] Work area for monitor 0: 2560 32 2560x1408
Nov 17 23:56:40 efog-laptop gnome-shell[320243]: [tilingshell] [TilingManager 1] Work area for monitor 1: 0 0 2560x1440
Nov 17 23:56:40 efog-laptop gnome-shell[320243]: [tilingshell] [extension] extension is enabled
Nov 17 23:56:40 efog-laptop gnome-shell[320243]: [tilingshell] [DefaultMenu] [{"name":"Samsung Electric Company 27\"","x":2560,"y":0,"width":2560,"height":1440},{"name":"Samsung Electric Company 27\"","x":0,"y":0,"width":2560,"height":1440}]
Nov 17 23:56:41 efog-laptop gnome-shell[320243]: [tilingshell] [TilingManager 0] new work area for monitor 0: 0 0 2560x1440
Nov 17 23:56:41 efog-laptop gnome-shell[320243]: [tilingshell] [TilingManager 1] new work area for monitor 1: 2560 32 2560x1408
```

![image](https://github.com/user-attachments/assets/9f4d1163-a0f3-4629-98c7-99d52d828680)

**Comment by efogdev** [comment]
(somehow fixed after several reboots)

**Comment by efogdev** [comment]
Reopening because it randomly reproduces on monitors connect

**Comment by domferr** [comment]
Hey, thank you for sharing this. I noticed this bug may appear when switching monitors in some particular situations and setups. Indeed, normally when a monitor gets added or removed everything works smooth. To my experience disabling and enabling Tiling Shell doesn't fix the problem and I assumed it is a GNOME shell bug. Indeed, after a reboot the bug is no longer there.

**Comment by efogdev** [comment]
Yeah, unfortunately GNOME + multi-monitor is a mess on laptops :C 
Let me know if you need any debug information.

**Comment by efogdev** [comment]
Also it wasn't a thing for me before v15

**Comment by bmaupin** [comment]
I am experiencing something similar. I'm guessing it's the same bug?

If I drag a window to another monitor, it will highlight the wrong monitor. In my case the window doesn't fill the whole screen, it will snap to a side or a corner. But it will always be the wrong monitor.

I'm hoping this will be fixed by https://github.com/domferr/tilingshell/issues/197 🤞

One other thing in case it's helpful: I don't notice this bug at all in X11, only in Wayland.

**Comment by domferr** [comment]
Thank you for your feedback guys. Do you have fractional scaling enabled?

**Comment by bmaupin** [comment]
I have fractional scaling enabled on one of the monitors but not the other.

**Comment by bmaupin** [comment]
@efogdev @domferr I manually updated to the next version (15.1) from here: https://github.com/domferr/tilingshell/issues/197#issuecomment-2501245226

It seems to fix this bug. So maybe that means this is a duplicate of #197 ?

**Comment by ccoenen** [comment]
I run into this frequently, too. This is a Fedora 41 / Gnome 47 / Wayland setup with no fractional scaling. Both monitors are at 100%. For some reason, my computer sometimes forgets how I set up my monitors and flips them around. After putting them back in the correct layout, Tiling Shell will not pick up on that change. And I get pretty much what @domferr deescribes, with the only difference being that Disabling and re-enabling Tiling Shell will fix my issue (seen from ca. 0:30 in that video).

https://github.com/user-attachments/assets/471acc71-c047-4ee9-8575-60cd534e2356

**Comment by bmaupin** [comment]
The workaround of disabling and re-enabling the extension is working for me too.

@domferr Are there any logs that we can collect to help troubleshoot this issue? Thanks!

**Comment by domferr** [comment]
Hey, I'm looking into this, however currently I'm not in my apartment and I haven't a secondary monitor atm. I'm then _guessing_ a solution. Let me know if it solves the problem! Thank you for your help :heart: 

- GNOME 42, 43 or 44: [GNOME.42-44.tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18028917/GNOME.42-44.tilingshell%40ferrarodomenico.com.zip)
- GNOME 45, 46 or 47: [tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18028919/tilingshell%40ferrarodomenico.com.zip)

**Comment by domferr** [comment]
Hey everyone, can you confirm this bug doesn't appear anymore with Tiling Shell 15.1? Thank you

**Comment by bmaupin** [comment]
I'm using 15.1 and I haven't seen it yet. I'll try to follow up again in a week or so.

**Comment by efogdev** [comment]
Same!

**Comment by bmaupin** [comment]
I wasn't getting the bug with Wayland but I actually switched back to X11 to troubleshoot an Ubuntu bug (https://bugs.launchpad.net/ubuntu/+source/gnome-shell/+bug/2089709) and now I'm getting this bug again.

I think I'm using the latest version. In Gnome Extension Manager it says version 43.

**Comment by ccoenen** [comment]
I am on 15.1 and I believe I'm still running into the issue. Reproduction are these steps:

1. have two monitors, have tilingshell installed
2. verify that tilingshell is working correctly
3. go to your system's settings, screen settings and change the layout of your two monitors
4. after that change, the issue like in my video above will appear again.

**Comment by domferr** [comment]
Thank you guys for sharing your feedback! I'm happy that the bug is overall fixed but that there are only some edge cases not working properly. I'll look into them

**Comment by efogdev** [comment]
It's back :'( 
Btw my `.config/monitors.xml` is a big mess, and GNOME always does that

**Comment by domferr** [comment]
Yeah, I'll give you a bit of context about this bug. Tiling Shell looks at the mouse position to understand on which monitor the mouse is. According to that, the tiles on that monitor are shown. The problem here is that for some reasons GNOME provides wrong monitors coordinates. That happens sometimes and only when GNOME struggles on monitor setup changes, scaling factor changes, etc...
I'm researching more about this

**Comment by domferr** [comment]
I just found out the method [`get_monitor_index_for_rect`](https://mutter.gnome.org/meta/method.Display.get_monitor_index_for_rect.html). Hope it is more reliable! If you can, give it a try, thank you! :heart: 

- GNOME 42, 43 or 44: [GNOME.42-44.tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18247231/GNOME.42-44.tilingshell%40ferrarodomenico.com.zip)
- GNOME 45, 46 or 47: [tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18247232/tilingshell%40ferrarodomenico.com.zip)

**Comment by efogdev** [comment]
Installed it, will report in a few days

**Comment by bmaupin** [comment]
>     * GNOME 45, 46 or 47: [tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18247232/tilingshell%40ferrarodomenico.com.zip)

I've been testing this on Gnome 46 and X11 since January 10. No issues so far. I'll wait a little bit longer and I'll test with Wayland.

I think this also fixed another issue I had, where sometimes after logging in I would only get a black screen with an X mouse cursor.

**Comment by domferr** [comment]
Hey everyone! Do you think we can close this? Can you confirm the problem is solved for you?

**Comment by ccoenen** [comment]
The problem I showed in this https://github.com/domferr/tilingshell/issues/194#issuecomment-2507261741 is fully resolved in Version 16.1. Thank you!

**Comment by efogdev** [comment]
Hey! I believe this is resolved, thank you!

**Comment by bmaupin** [comment]
I've had no issues with X11. I haven't been able to test with Wayland but I would expect everyone else is using it so I'm guessing it's fixed.

**Comment by domferr** [comment]
Nice! Thank you everybody 💟
